sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -195,6 +196,35 @@ trait HadoopFsRelationProvider {
     }
     createRelation(sqlContext, paths, dataSchema, partitionColumns, parameters)
   }
+
+  override def createSource(
+      sqlContext: SQLContext,
+      schema: Option[StructType],
+      providerName: String,
+      parameters: Map[String, String]): Source = {
+    val path = parameters("path")
+    val metadataPath = parameters.getOrElse("metadataPath", s"$path/_metadata")
+
+    def dataFrameBuilder(files: Array[String]): DataFrame = {
+      val relation = createRelation(
+        sqlContext,
+        files,
+        schema,
+        partitionColumns = None,
+        bucketSpec = None,
+        parameters)
+      DataFrame(sqlContext, LogicalRelation(relation))
+    }
+
+    val maybeSchema =
+      if (schema.isEmpty && providerName == "text") {
```

Comment:
My point was that I don't want to special case the logic here, but should instead defer to the source.  What happens if we add another source that is capable of producing a schema without any files present?